### PR TITLE
SONARKT-98 S6309: Functions returning Flow/Channel should not be suspending

### PR DESCRIPTION
**Pull request by johann-beleites-sonarsource**

### CODE REVIEW

File: sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/api/ApiExtensions.kt [hunk 1]
```diff
@@ -252,3 +253,9 @@ private fun KtParameter.determineType(bindingContext: BindingContext) =
 
 fun KtParameter.determineTypeAsString(bindingContext: BindingContext, printTypeArguments: Boolean = false) =
     determineType(bindingContext)?.getJetTypeFqName(printTypeArguments)
+
+fun KtNamedFunction.returnTypeAsString(bindingContext: BindingContext, printTypeArguments: Boolean = false) =
+    returnType(bindingContext)?.getJetTypeFqName(printTypeArguments)
+
+fun KtNamedFunction.returnType(bindingContext: BindingContext) =
```
Comment (margarita-nedzelska-sonarsource) [inline sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/api/ApiExtensions.kt:260]: It looks like this function could be private. Also. maybe it makes sense to combine this function with `returnTypeAsString`?

File: sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/FlowChannelReturningFunsNotSuspendingCheck.kt [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+import org.sonarsource.kotlin.plugin.KotlinFileContext
+
+private val FORBIDDEN_RETURN_TYPES = listOf(COROUTINES_FLOW, COROUTINES_CHANNEL)
+private const val MESSAGE = """Functions returning "Flow" or "Channel" should not be suspending"""
+
+@Rule(key = "S6309")
+class FlowChannelReturningFunsNotSuspendingCheck : AbstractCheck() {
+    override fun visitNamedFunction(function: KtNamedFunction, kotlinFileContext: KotlinFileContext) {
+        val suspendModifier = function.suspendModifier()
+        if (suspendModifier != null && function.returnTypeAsString(kotlinFileContext.bindingContext) in FORBIDDEN_RETURN_TYPES) {
```
Comment (margarita-nedzelska-sonarsource) [inline sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/FlowChannelReturningFunsNotSuspendingCheck.kt:21]: How about `function.suspendModifier()?.let {...}`?

File: sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/FlowChannelReturningFunsNotSuspendingCheck.kt [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+            val secondaries = function.typeReference
+                ?.let { SecondaryLocation(kotlinFileContext.textRange(it)) }
+                ?.let { listOf(it) }
```
Comment (margarita-nedzelska-sonarsource) [inline sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/FlowChannelReturningFunsNotSuspendingCheck.kt:24]:
```suggestion
                ?.let { listOf(SecondaryLocation(kotlinFileContext.textRange(it))) }
```

**Review by margarita-nedzelska-sonarsource — APPROVED** [review]
Overall looks good, few minor suggestions

File: sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/api/ApiExtensions.kt (hunk 1 above)
Comment (johann-beleites-sonarsource) [reply sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/api/ApiExtensions.kt:260]: My idea behind this was that you may want to perform some further processing on the type, for which you might need the actual `KotlinType`. So by splitting it up like this, you still have that option in the API. WDYT?

**Pull request merged**